### PR TITLE
Add the locations to the field in an intersected record

### LINKED ISSUE #42094: [Bug]: `RecordTypeSymbol.fieldDescriptors()` returns NPE
### Description

I have an error intersection like this:
```bal

type ErrorDetail record {
    string|string[] headers?;
    anydata body?;
};
type StatusCodeError distinct error<ErrorDetail>;
type NotFoundError distinct StatusCodeError;

type UserErrorInfo record {|
    string timeStamp;
    string userId;
    string message;
|};

type UserErrorDetail record {|
    *ErrorDetail;
    UserErrorInfo body;
|};

type Error distinct error;

type UserNotFoundError Error & NotFoundError & error<UserErrorDetail>;
```

In the compiler plugin, I am accessing this `UserNotFoundError` type symbol and get the effective error type from the intersection. Then I get the type descriptor of the detail field which is a Ballerina Record Type in this case.

When I call `fieldDescriptors()` method on this Record Type I am getting a Null Pointer Error.

Screenshot 2024-02-01 at 18 30 14

**Please note that this a blocker to this feature which is expected to deliver with update 9:** 
- https://github.com/ballerina-platform/ballerina-library/issues/5135

### Steps to Reproduce

See the description above.

Java code:
```java
TypeSymbol intersectionErrorType = ((TypeReferenceTypeSymbol) typeSymbol).typeDescriptor();
TypeSymbol effectiveErrorType = ((IntersectionTypeSymbol) intersectionErrorType).effectiveTypeDescriptor();
TypeSymbol detailRecordType = ((ErrorTypeSymbol) effectiveErrorType).detailTypeDescriptor();
((RecordTypeSymbol) detailRecordType).fieldDescriptors();
```

> **Note:**
> If I change the error intersection like this then it works:
> ```ballerina
> type Error0 distinct error<UserErrorDetail>;
> 
> type UserNotFoundError0 Error0 & NotFoundError;
> ```

### Affected Version(s)

Ballerina SwanLake Update 8 (2201.8.4)

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by nipunayf**
## Purpose
The current implementation sets `null` as the location when adding the fields to the new intersected record. This impacts the `fieldDescriptors` API as it is unable to detect the line range when initializing the symbols.

Fixes #42094

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/symbols/error_type_symbol_test.bal [hunk 1]
```diff
@@ -43,3 +43,23 @@ function test() {
...
+
+type SimpleError error<ErrorDetail1>;
+type InclusionError error<ErrorDetail2>;
+
+type SimpleIntersectionError distinct error & error<ErrorDetail1> & error<ErrorDetail1>;
+type MultipleIntersectionError InclusionError & error<ErrorDetail1> & SimpleError & error<ErrorDetail2>;
+function testMultipleIntersectionError() {
+    SimpleIntersectionError simpleErr;
+    MultipleIntersectionError multipleErr;
+}
```
Comment (SasinduDilshara) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/symbols/error_type_symbol_test.bal:65]: in the original issue, there are union types defined in the error type record details.
shall we add a one test for that scenario?

**Review by LakshanWeerasinghe — DISMISSED** [review]

**Review by KavinduZoysa — DISMISSED** [review]

**Comment by KavinduZoysa** [comment]
@MaryamZi, Do we need to send this fix to 8.x branch?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -5618,7 +5618,7 @@ private boolean populateFields(IntersectionContext intersectionContext, BRecordT
                                                    intersectionFieldType, newTypeSymbol, lhsRecordField.pos, SOURCE);
             }
 
-            newTypeFields.put(key, new BField(name, null, recordFieldSymbol));
+            newTypeFields.put(key, new BField(name, recordFieldSymbol.pos, recordFieldSymbol));
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:5621]: It is correct to set the position here. My concern is why we keep the position as a field of `BField`. Normally, we access the position from the symbol. Here as well, when we need to retrieve the position, we need to obtain it from the symbol.

Comment (nipunayf) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:5621]: When we create the record field symbol of the resultant record symbol of the intersection, we use the position of this `BField`. Alternatively, we can use `lhsRecordField.symbol.pos` as well here.

https://github.com/ballerina-platform/ballerina-lang/blob/9e6acd237f01eb4575e69f137c4f1206ef6ff8f3/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java#L5619

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:5621]: Yes. IMO, That is the best way to access the position. Shall we try whether we can remove `pos` in the `BField` without affecting the code?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 3]
```diff
@@ -5615,10 +5615,10 @@ private boolean populateFields(IntersectionContext intersectionContext, BRecordT
                 invokableSymbol.flags = tsymbol.flags;
             } else {
                 recordFieldSymbol = new BVarSymbol(intersectionFlags, name, env.enclPkg.packageID,
-                                                   intersectionFieldType, newTypeSymbol, lhsRecordField.pos, SOURCE);
+                        intersectionFieldType, newTypeSymbol, lhsRecordField.symbol.pos, SOURCE);
             }
 
-            newTypeFields.put(key, new BField(name, recordFieldSymbol.pos, recordFieldSymbol));
+            newTypeFields.put(key, new BField(name, null, recordFieldSymbol));
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:5621]: I think you got what I said wrong. What I meant was, whether we can remove the `pos` field from the `BField`. If we can't, then let's just use `recordFieldSymbol.pos` for `pos` here.

Comment (nipunayf) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Currently, there are multiple usages that use the `pos` field of the `BField`. We may require a separate issue to evaluate and refactor this change if applicable.

**Review by KavinduZoysa — APPROVED** [review]

**Pull request merged**